### DISCUSSION TITLE
Creating a custom CP module

### DISCUSSION THREAD

**Discussion by ehbfa** [statamic/cms#8562] [Troubleshooting]
Heya! I want to write a custom module to handle incoming registrations (form submissions). Because I need to do all kinds of extra stuff with that data, like grouping it based on the event (even though there is only one form), and add extra features like 'download PDF'.

Now I have done this in the past for an older project, but this time, when I use that code - it fails to run. So I'm trying to go back to the basics, and see why/how I should set this up in newer statamic projects. The old project was on Statamic/cms 3, as is the new one, but the old project was using webpack and the new one is using vite. And when I try to compile the code, it fails on the Vue templates. But I cannot figure out how to solve that.

So my question is; what is a good approach to create a custom CP module, with a table and detail view? 

A bit more off topic (as that is probably more a vite question), is why does my compilation fail with an error pointing to my template code which runs fine in webpack. I can of course provide the error, if this isn't too much off topic :)

**Comment by edalzell** [comment]
What do you mean "fails to run"? Should all still work, V4 didn't change that many things.

You make CP modules the same way as before, in Vue 2 (usually), then compile them into a js file that's loaded by Statamic.

Take a look at Runway, for example, it has custom CP pages.

**Reply by ehbfa** [reply]
I just did install Runway to test it out, but it breaks on me for not having the file **HasRunwayResource** available nowhere, which I should use as a trait. I tried to use it, but it just isn't there :)

---

And for my Vue issue; I've added `resources/js/cp.js` to my list of inputs in my `vite.config.js`. And in that cp.js file, there is only; `import RegistrationsListing from "./components/registrations/Listing.vue"` And in that Listing.vue file, the contents is like this;

```vue
<template><div></div></template>

<script>
import Listing from '../../../../vendor/statamic/cms/resources/js/components/Listing.vue'

export default {

  mixins: [Listing],

  props: {
    registrationsRequestUrl: '',
    editRoute: '',
  },

  data() {
    return {
      rows: this.initialRows,
      requestUrl: this.registrationsRequestUrl,
      preferencesPrefix: 'registrations',
    }
  },

  methods: {
    getStatusClass(entry) {
      if (entry.published && entry.private) {
        return 'bg-transparent border border-grey-60';
      } else if (entry.published) {
        return 'bg-green';
      } else {
        return 'bg-grey-40';
      }
    },
  },

}
</script>
```

(had to put this content out of the code block, or the template part would be ignored)

And when I run sail npm run production, I get this error;

```
[vite:build-import-analysis] Parse error @:3:9
file: /var/www/html/resources/js/components/registrations/Listing.vue:3:8
1: <template>
2:   <div>
3:   </div>
           ^
4: </template>
error during build:
Error: Parse error @:3:9
    at parse$e (file:///var/www/html/node_modules/vite/dist/node/chunks/dep-8609dc5d.js:16478:355)
    at Object.transform (file:///var/www/html/node_modules/vite/dist/node/chunks/dep-8609dc5d.js:46472:27)
```

**Reply by duncanmcclean** [reply]
What does your package.json file look like? Are you using Vue 2 or Vue 3?

(I've tidied up the code formatting in your comment for you - using three backticks around code blocks instead of one)
